unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1427 @@
+        )
+    }
+
+    #[test]
+    fn test_scheduling_state_machine_creation() {
+        let state_machine = unsafe {
+            SchedulingStateMachine::exclusively_initialize_current_thread_for_scheduling()
+        };
+        assert_eq!(state_machine.active_task_count(), 0);
+        assert_eq!(state_machine.total_task_count(), 0);
+        assert!(state_machine.has_no_active_task());
+    }
+
+    #[test]
+    fn test_scheduling_state_machine_reinitialization() {
+        let mut state_machine = unsafe {
+            SchedulingStateMachine::exclusively_initialize_current_thread_for_scheduling()
+        };
+        state_machine.total_task_count.increment_self();
+        assert_eq!(state_machine.total_task_count(), 1);
+        state_machine.last_task_index = Some(1);
+        state_machine.reinitialize();
+        assert_eq!(state_machine.total_task_count(), 0);
+        assert_eq!(state_machine.last_task_index, None);
+    }
+
+    #[test]
+    fn test_create_task() {
+        let sanitized = simplest_transaction();
+        let task = SchedulingStateMachine::create_task(sanitized.clone(), 3, &mut |_| {
+            UsageQueue::default()
+        });
+        assert_eq!(task.task_index(), 3);
+        assert_eq!(task.transaction(), &sanitized);
+    }
+
+    #[test]
+    fn test_non_conflicting_task_related_counts() {
+        let sanitized = simplest_transaction();
+        let address_loader = &mut create_address_loader(None);
+        let task = SchedulingStateMachine::create_task(sanitized.clone(), 3, address_loader);
+
+        let mut state_machine = unsafe {
+            SchedulingStateMachine::exclusively_initialize_current_thread_for_scheduling()
+        };
+        let task = state_machine.schedule_task(task).unwrap();
+        assert_eq!(state_machine.active_task_count(), 1);
+        assert_eq!(state_machine.total_task_count(), 1);
+        state_machine.deschedule_task(&task);
+        assert_eq!(state_machine.active_task_count(), 0);
+        assert_eq!(state_machine.total_task_count(), 1);
+        assert!(state_machine.has_no_active_task());
+    }
+
+    #[test]
+    fn test_conflicting_task_related_counts() {
+        let sanitized = simplest_transaction();
+        let address_loader = &mut create_address_loader(None);
+        let task1 = SchedulingStateMachine::create_task(sanitized.clone(), 101, address_loader);
+        let task2 = SchedulingStateMachine::create_task(sanitized.clone(), 102, address_loader);
+        let task3 = SchedulingStateMachine::create_task(sanitized.clone(), 103, address_loader);
+
+        let mut state_machine = unsafe {
+            SchedulingStateMachine::exclusively_initialize_current_thread_for_scheduling()
+        };
+        assert_matches!(
+            state_machine
+                .schedule_task(task1.clone())
+                .map(|t| t.task_index()),
+            Some(101)
+        );
+        assert_matches!(state_machine.schedule_task(task2.clone()), None);
```

Comment:
> Seeing these tests has made me somewhat concerned on the use of `index` within the state machine.
> The scheduler will schedule **not** based on index, but on the order in which transactions are fed in. It seems like the index is only tracked for sanity checks, but it feels like a layer violation

wow (again), quite good observation. Yeah, I agree. In retrospect, I added these sanity checks too prematurely, just because i can. I'll rethink on this later. for now, I just removed the assertions: d072efdd4d721613bd6ea89f96e3af060aa92888

(Lastly, I think you commented on wrong lines..)

Reply:
(this review thread is referenced by https://github.com/anza-xyz/agave/pull/3058#discussion_r1787839215)